core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -88,33 +84,30 @@ public class HTTPClient implements RESTClient {
 
   @VisibleForTesting static final String REST_SOCKET_TIMEOUT_MS = "rest.client.socket-timeout-ms";
 
-  private final String uri;
+  private final URI baseUri;
   private final CloseableHttpClient httpClient;
+  private final Map<String, String> baseHeaders;
   private final ObjectMapper mapper;
+  private final AuthSession authSession;
 
   private HTTPClient(
-      String uri,
+      URI baseUri,
       HttpHost proxy,
       CredentialsProvider proxyCredsProvider,
       Map<String, String> baseHeaders,
       ObjectMapper objectMapper,
       HttpRequestInterceptor requestInterceptor,
       Map<String, String> properties,
       HttpClientConnectionManager connectionManager) {
-    this.uri = uri;
+    this.baseUri = baseUri;
+    this.baseHeaders = baseHeaders;
     this.mapper = objectMapper;
+    this.authSession = AuthSession.EMPTY;
```

Comment:
I think defaulting to empty opens up a risk that requests will be made without an explicit session.  This somewhat hides the case where we missed a call that should have auth.  I would think we would leave this a null and require that a session be used (e.g. validate the auth was set in the `execute` call).  Thoughts @nastra?

Reply:
I think that makes sense to enforce/validate that a proper auth session is being used

Reply:
We are going to run into a chicken and egg problem: to create an auth session I need an HTTP client; but now, to create an HTTP client, I would need an auth session.

Reply:
I understand that may seem like the case here, but I think we're just saying that you create it initially as set to `null`, but validate when `execute` is called that it is not null.  You can still create the HTTPClient, but you can't make a request unless there is an active auth session.  This just prevents unintentional use of an "empty" auth session (it needs to be intentional).

Reply:
Ah, now I get it. Let me try your suggestion.